yarn-project/simulator/src/public/benchmarker.ts

```diff
@@ -0,0 +1,54 @@
+import { type L2BlockBuiltStats } from '@aztec/circuit-types/stats';
+import { elapsed } from '@aztec/foundation/timer';
+import { type ToCamelCase } from '@aztec/foundation/types';
+
+import { type PublicKernelPhase } from './abstract_phase_manager.js';
+
+type Event = `kernel-simulation` | `avm-simulation`;
+type PhaseEvent = `${PublicKernelPhase}-${Event}`;
+type PublicEventDurations = Record<PhaseEvent, number[]>;
+
+// exported benchmarks need to be in camel case
+type PhaseEventBenchmark = `${ToCamelCase<PhaseEvent>}AvgMs`;
+
+function camelCasePhaseEvent(phaseEvent: PhaseEvent): PhaseEventBenchmark {
+  const camelCased = phaseEvent.replace(/-([a-z])/g, g => g[1].toUpperCase());
+  return `${camelCased}AvgMs` as PhaseEventBenchmark;
+}
+
+type AverageDurations = Record<PhaseEventBenchmark, number>;
+
+export class PublicBenchmarker {
+  public durations: PublicEventDurations;
+
+  constructor() {
+    this.durations = {
+      'setup-kernel-simulation': [],
+      'setup-avm-simulation': [],
+      'app-logic-kernel-simulation': [],
+      'app-logic-avm-simulation': [],
+      'teardown-kernel-simulation': [],
+      'teardown-avm-simulation': [],
+      'tail-kernel-simulation': [],
+      'tail-avm-simulation': [],
+    };
+  }
+
+  async benchmark<T>(phase: PublicKernelPhase, event: Event, callback: () => T) {
+    const [durationMs, result] = await elapsed(callback);
+    this.durations[`${phase}-${event}`].push(durationMs);
```

Comment:
Highly recommend the new histogram in Node

https://nodejs.org/dist/latest-v18.x/docs/api/perf_hooks.html#perf_hookscreatehistogramoptions